rust_icu_ubrk/src/lib.rs

```diff
@@ -0,0 +1,866 @@
+                Some(loc)
+            }
+            _ => None,
+        }
+    }
+}
+
+impl ExactSizeIterator for Locales {
+    /// Reports the number of locales for which text breaking information is available.
+    ///
+    /// Implements `ubrk_countAvailable`.
+    fn len(&self) -> usize {
+        self.upper as usize
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::UBreakIterator;
+    use log::trace;
+    use rust_icu_sys::{self as sys, UBreakIteratorType::*, ULocDataLocaleType::*};
+    use std::{convert::TryFrom, rc::Rc};
+
+    // Wraps a `UBreakIterator` to emit Strings formed by pairs of word boundaries.
+    struct Words<'a> {
+        iter: &'a mut UBreakIterator,
+        chars: Vec<sys::UChar>,
+    }
+
+    impl<'a> Words<'a> {
+        fn new(iter: &'a mut UBreakIterator) -> Self {
+            let text = String::try_from(&*iter.text).unwrap();
+            Self {
+                iter,
+                chars: text.as_str().encode_utf16().collect::<Vec<_>>(),
+            }
+        }
+    }
+
+    impl<'a> Iterator for Words<'a> {
+        type Item = String;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            let start = self.iter.current();
+            self.iter.next().and_then(|end| {
+                String::from_utf16(
+                    &self.chars[(start as usize)..(end as usize)],
+                )
+                .ok()
+            })
+        }
+    }
+
+    const TEXT: &str =
+        r#""It wasn't the wine," murmured Mr. Snodgrass. "It was the salmon.""#;
+
+    #[test]
+    fn test_iteration() {
+        let mut break_iter =
+            UBreakIterator::try_new(UBRK_WORD, "en", TEXT).unwrap();
+
+        assert!(break_iter.is_boundary(0));
+        assert_eq!(0, break_iter.first());
+        assert_eq!(None, break_iter.previous());
+        assert_eq!(0, break_iter.current());
+
+        let word_start = TEXT.find("murmured").unwrap() as i32;
+        let word_end = word_start + 8;
+        assert!(break_iter.is_boundary(word_start));
+        assert!(break_iter.is_boundary(word_end));
+        assert!(!break_iter.is_boundary(word_start + 3));
+        assert_eq!(word_end, break_iter.following(word_start + 3));
+        assert_eq!(word_end, break_iter.current());
+        assert_eq!(Some(word_start), break_iter.previous());
+        assert_eq!(word_start, break_iter.current());
+        assert_eq!(Some(word_end), break_iter.next());
+        assert_eq!(word_end, break_iter.current());
+        assert_eq!(word_start, break_iter.preceding(word_start + 3));
+        assert_eq!(word_start, break_iter.current());
+
+        let text_len = TEXT.len() as i32;
+        assert!(break_iter.is_boundary(text_len));
+        assert_eq!(text_len, break_iter.last_boundary());
+        assert_eq!(None, break_iter.next());
+        assert_eq!(text_len, break_iter.current());
+
+        break_iter.first();
+        let word_iter = Words::new(&mut break_iter);
+        assert_eq!(
```

Comment:
Thanks for making this change.  Is there a specific reason that `word_iter` takes `break_iter`as mutable?

Reply:
Calling `next()` on an iterator requires a mutable reference, and since `Words` calls `next()` on the break iterator, it needs to hold a mutable reference. Alternatively it could fully take ownership of the break iterator, since this is just a helper type for tests.